org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/DefaultGradleBuild.java

```diff
@@ -58,16 +63,24 @@ public SynchronizationResult synchronize(IProgressMonitor monitor) {
     }
 
     public SynchronizationResult synchronize(NewProjectHandler newProjectHandler, CancellationTokenSource tokenSource, IProgressMonitor monitor) {
+        synchronizingBuilds.add(this);
```

Comment:
Is it possible that this method is called from different threads, e.g. when synchronization is requested synchronously after the job has just been scheduled? If so, should check the result of calling `add()` and abort the synchronization if one is already in progress?

Reply:
It's can happen but the other way around. The synchronization is scheduled synchronously, which triggers Eclipse JDT to initialize the classpath container with a new sync job.
It makes sense, although I just realized that the proper solution might be to implement scheduling rules. I'm going to look into that tomorrow.

Reply:
The scheduling rule solution was a big dead-end. I had the assumption that I can amend existing scheduling rules which turned out to be not true as there's no way to extend the workspace root rule with new functionality (`GradleBuildRule`).

I managed to solve the situation with your proposal: the synchronize operations are skipped within `DefaultGradleBuild` if there's already running on a different thread.